### DISCUSSION TITLE
EVP_aes_256_ctr pkcs7 padding

### DISCUSSION THREAD

**Discussion by vaassi** [openssl/openssl#21888] [Q&A] [answered]
Hi, there!

I have issue with padding during decryption. Docs tells that padding is applied by default, but in my case seems that padding for some reason not working. May be someone have ideas why this code don't work as expected. Thanks.

My code for decryption:

    using EVP_CIPHER_CTX_ptr = std::unique_ptr<EVP_CIPHER_CTX, decltype(&::EVP_CIPHER_CTX_free)>;

    std::vector<uint8_t> out(body.size() + IV_LENGTH);
    int out_len{};
    EVP_CIPHER_CTX_ptr ctx(EVP_CIPHER_CTX_new(), ::EVP_CIPHER_CTX_free);
    EVP_DecryptInit_ex(ctx.get(), EVP_aes_256_ctr(), nullptr, key1.data(), iv.data());
    EVP_CIPHER_CTX_set_padding(ctx.get(), EVP_PADDING_PKCS7);
    EVP_DecryptUpdate(ctx.get(), out.data(), &out_len, body.data(), static_cast<int>(body.size()));
    size_t total = out_len;

    EVP_DecryptFinal_ex(ctx.get(), out.data() + out_len, &out_len);
    out.resize(total + out_len);

    std::cout << "Decrypted data:" << std::endl;
    BIO_dump_fp(stdout, out.data(), static_cast<int>(out.size()));

if I try to dump out result I get:
`
Decrypted data:
0000 - 31 32 33 0a 0c 0c 0c 0c-0c 0c 0c 0c 0c 0c 0c 0c   123.............
`

right now I do manual unpad, but seems that this solution is not perfect:

    if (out[out.size() - 1] <= IV_LENGTH) {
        auto end = std::find_if(out.rbegin(), out.rend(), [](auto i) { return i == 0xa; });
        out.resize(std::distance(out.begin(), end.base()));
    }

**Comment by mattcaswell** [comment, marked as answer]
> EVP_aes_256_ctr

CTR mode does not use padding and it makes no sense to have it. CTR encrypted data can be any arbitrary length and does not need to be a multiple of the underlying block size - so padding is not needed.

It seems to me that what is at fault here is the encryption process. It should not be adding padding for this cipher.

**Reply by vaassi** [reply]
Thank you for reply. Padding added by external tool. I only try to decrypt and use content.
Do openssl have tools for manually control padding process, I can't find any docs regarding this topic, or if padding is used, the only way to unpad is to do it manually?

**Reply by mattcaswell** [reply]
There is no mechanism provided by OpenSSL to remove padding that is incorrectly present.
